File: api-service/src/metrics/prometheus/metrics.ts
import Prometheus from "prom-client";

// Create a new Prometheus Gauge for query response time
const queryResponseTimeMetric = new Prometheus.Gauge({
    name: "node_query_response_time",
    help: "The average response time for database queries",
    labelNames: ["entity", "id", "endpoint", "dataset_id", "status", "request_size", "response_size"]
});

// Create a new Prometheus Counter for total API calls
const totalApiCallsMetric = new Prometheus.Counter({
    name: "node_total_api_calls",
    help: "The total number of API calls made",
    labelNames: ["entity", "id", "endpoint", "dataset_id", "status", "request_size", "response_size"]
});

// Create a new Prometheus Counter for failed API calls
const failedApiCallsMetric = new Prometheus.Counter({
    name: "node_failed_api_calls",
    help: "The total number of Failed API calls made",
    labelNames: ["entity", "id", "endpoint", "dataset_id", "status", "request_size", "response_size"]
});

// Create a new Prometheus Counter for success API calls
const successApiCallsMetric = new Prometheus.Counter({
    name: "node_success_api_calls",
    help: "The total number of Successful API calls made",
    labelNames: ["entity", "id", "endpoint", "dataset_id", "status", "request_size", "response_size"]
})

export {
    queryResponseTimeMetric,
    totalApiCallsMetric,
    failedApiCallsMetric,
    successApiCallsMetric
}

File: command-service/src/metrics/metrics.py
from prometheus_client import Counter, Gauge


class Metrics:
    def __init__(self, registry) -> None:
        self.node_query_response_time = Gauge(
            name="node_query_response_time",
            documentation="The average response time for database queries",
            labelnames=["entity", "id", "endpoint", "datasetId", "status"],
            registry=registry,
        )
        self.total_api_calls = Counter(
            name="node_total_api_calls",
            documentation="The total number of API calls made",
            labelnames=["entity", "id", "endpoint", "datasetId"],
            registry=registry,
        )
        self.failed_api_calls = Counter(
            name="node_failed_api_calls",
            documentation="The total number of failed API calls made",
            labelnames=["entity", "id", "endpoint", "datasetId", "status"],
            registry=registry,
        )
        self.success_api_calls = Counter(
            name="node_success_api_calls",
            documentation="The total number of successful API calls made",
            labelnames=["entity", "id", "endpoint", "datasetId", "status"],
            registry=registry,
        )

    def queryResponseTimeMetric(self):
        return self.node_query_response_time

    def totalApiCallsMetric(self):
        return self.total_api_calls

    def failedApiCallsMetric(self):
        return self.failed_api_calls

    def successApiCallsMetric(self):
        return self.success_api_calls
